porch/controllers/packagevariants/pkg/controllers/packagevariant/injection.go

```diff
@@ -0,0 +1,425 @@
+	var kptfile kptfilev1.KptFile
+	err := kptfileKubeObject.As(&kptfile)
+	if err != nil {
+		return err
+	}
+
+	info := kptfile.Info
+	if info == nil {
+		info = &kptfilev1.PackageInfo{}
+	}
+	// generate a unique list of gates (append is not idempotent)
+	gateMap := make(map[string]bool)
+	for _, gate := range info.ReadinessGates {
+		gateMap[gate.ConditionType] = true
+	}
+
+	status := kptfile.Status
+	if status == nil {
+		status = &kptfilev1.Status{}
+	}
+	conditions := convertConditionsToMeta(status.Conditions)
+	// set a condition for each injection point
+	for _, ip := range injectionPoints {
+		if ip.required {
+			gateMap[ip.conditionType] = true
+		}
+		var condStatus metav1.ConditionStatus
+		condStatus = "False"
+		condReason := "NoResourceSelected"
+		condMessage := "no resource matched any injection selector for this injection point"
+		if len(ip.errors) > 0 {
+			condMessage = strings.Join(ip.errors, ", ")
+		}
+		if ip.injected {
+			condStatus = "True"
+			condReason = "ConfigInjected"
+			condMessage = fmt.Sprintf("injected resource %q from cluster", ip.injectedName)
+		}
+
+		meta.SetStatusCondition(&conditions, metav1.Condition{
+			Type:    ip.conditionType,
+			Status:  condStatus,
+			Reason:  condReason,
+			Message: condMessage,
+		})
+
+	}
+
+	// update the readiness gates
+	// TODO: this loses comments right now, fix that
+	var gates []kptfilev1.ReadinessGate
+	for k := range gateMap {
+		gates = append(gates, kptfilev1.ReadinessGate{ConditionType: k})
+	}
+	sort.Slice(gates, func(i, j int) bool { return gates[i].ConditionType < gates[j].ConditionType })
+
+	if gates != nil {
+		info.ReadinessGates = gates
+		err = kptfileKubeObject.SetNestedField(info, "info")
+		if err != nil {
+			return err
+		}
+	}
+
+	// update the status conditions
+	if conditions != nil {
+		sort.Slice(conditions, func(i, j int) bool { return conditions[i].Type < conditions[j].Type })
+		status.Conditions = convertConditionsFromMetaToKptfile(conditions)
+		err = kptfileKubeObject.SetNestedField(status, "status")
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertConditionsFromMetaToKptfile(conditions []metav1.Condition) []kptfilev1.Condition {
+	var result []kptfilev1.Condition
+	for _, c := range conditions {
+		result = append(result, kptfilev1.Condition{
+			Type:    c.Type,
+			Reason:  c.Reason,
+			Status:  kptfilev1.ConditionStatus(c.Status),
+			Message: c.Message,
+		})
+	}
+	return result
+}
+
+func convertConditionsToMeta(conditions []kptfilev1.Condition) []metav1.Condition {
+	var result []metav1.Condition
+	for _, c := range conditions {
+		result = append(result, metav1.Condition{
+			Type:    c.Type,
+			Reason:  c.Reason,
+			Status:  metav1.ConditionStatus(c.Status),
+			Message: c.Message,
+		})
+	}
+	return result
+}
+
+var matchResourceContents = append(kio.MatchAll, kptfilev1.KptFileName)
```

Comment:
could you add a comment explaining what this does?

Reply:
Yeah, this should move to a util function. We've copied it in a few places, I think.

Reply:
for now I added a comment and a TODO. There is some refactoring that can be done with all of this to make it nicer, this would be part of it. But I'd rather do that in a separate PR.